gtk/TorrentCellRenderer.cc

```diff
@@ -170,151 +143,161 @@ Glib::ustring getShortTransferString(
 
     case TR_STATUS_CHECK:
-        gstr += gtr_sprintf(_("Verifying local data (%.1f%% tested)"), tr_truncd(st->recheckProgress * 100.0, 1));
-        break;
+        return fmt::format(
+            _("Verifying local data ({percent_done:.1}% tested)"),
+            fmt::arg("percent_done", tr_truncd(st->recheckProgress * 100.0, 1)));
 
     case TR_STATUS_DOWNLOAD:
     case TR_STATUS_SEED:
-        {
-            /* download/upload speed, ratio */
-            gstr += gtr_sprintf("%s  ", getShortTransferString(tor, st, uploadSpeed_KBps, downloadSpeed_KBps));
-            gstr += gtr_sprintf(_("Ratio: %s"), tr_strlratio(st->ratio));
-            break;
-        }
+        return fmt::format(
+            "{} {}",
+            getShortTransferString(tor, st, uploadSpeed_KBps, downloadSpeed_KBps),
+            fmt::format(_("Ratio: {percent}"), fmt::arg("percent", tr_strlratio(st->ratio))));
 
     default:
+        return {};
+    }
+}
+
+static std::optional<std::string> getErrorString(tr_stat const* st)
+{
+    switch (st->error)
+    {
+    case TR_STAT_TRACKER_WARNING:
+        return fmt::format(_("Tracker warning: '{warning}'"), fmt::arg("warning", st->errorString));
         break;
+
+    case TR_STAT_TRACKER_ERROR:
+        return fmt::format(_("Error: '{error}'"), fmt::arg("error", st->errorString));
+        break;
+
+    case TR_STAT_LOCAL_ERROR:
+        return fmt::format(_("Local error: '{error}'"), fmt::arg("error", st->errorString));
+
+    default:
+        return std::nullopt;
     }
+}
 
-    return gstr;
+static auto getActivityString(
+    tr_torrent const* const tor,
+    tr_stat const* const st,
+    double const uploadSpeed_KBps,
+    double const downloadSpeed_KBps)
+{
+    switch (st->activity)
+    {
+    case TR_STATUS_STOPPED:
+    case TR_STATUS_CHECK_WAIT:
+    case TR_STATUS_CHECK:
+    case TR_STATUS_DOWNLOAD_WAIT:
+    case TR_STATUS_SEED_WAIT:
+        return getShortStatusString(tor, st, uploadSpeed_KBps, downloadSpeed_KBps);
+
+    case TR_STATUS_DOWNLOAD:
+        if (!tr_torrentHasMetadata(tor))
+        {
+            return fmt::format(
+                ngettext(
+                    "Downloading metadata from {active_count} connected peer ({percent_done:d}% done)",
+                    "Downloading metadata from {active_count} connected peers ({percent_done:d}% done)",
+                    st->peersConnected),
+                fmt::arg("active_count", st->peersConnected),
+                fmt::arg("percent_done", 100.0 * st->metadataPercentComplete));
+        }
+
+        if (st->peersSendingToUs != 0 && st->webseedsSendingToUs != 0)
+        {
+            return fmt::format(
+                ngettext(
+                    "Downloading from {active_count} of {connected_count} connected peer and webseed",
+                    "Downloading from {active_count} of {connected_count} connected peers and webseeds",
+                    st->peersSendingToUs + st->webseedsSendingToUs),
```

Comment:
Pluralized against `connected_count` (total), not `active_count` (part).
```suggestion
                    st->peersConnected + st->webseedsSendingToUs),
```

Reply:
67332b16e00c7f5f5e2ae8c9d70e5ed800e66b57